File: packages/swt3-ai-ts/src/adapters/agt.ts
/**
 * SWT3 AI Witness SDK -- Microsoft AGT (Agent Governance Toolkit) Adapter.
 *
 * Wraps any object with an evaluate() method (Microsoft AGT policy engine pattern),
 * minting witness anchors on each policy evaluation without modifying
 * the governance logic or adding framework dependencies.
 *
 * SWT3 is the independent witness layer for AGT-managed agents. AGT enforces
 * policy; SWT3 proves what AGT decided, cryptographically, out-of-band.
 *
 * Usage:
 *   import { wrapAGT } from "@tenova/swt3-ai/adapters/agt";
 *   const witnessed = wrapAGT(policyEngine, witness);
 *   const decision = await witnessed.evaluate(prompt, { model: "gpt-4o" });
 *
 * Duck-typed: works with any object that has an evaluate() method.
 * No Microsoft SDK import required.
 *
 * Copyright (c) 2026 Tenable Nova LLC. Apache 2.0. Patent pending.
 */

import type { Witness } from "../witness.js";
import type { InferenceRecord } from "../types.js";
import { sha256Truncated } from "../fingerprint.js";

export interface AGTEngine {
  evaluate(prompt: unknown, ...args: unknown[]): unknown;
  assess?(config: unknown, ...args: unknown[]): unknown;
  name?: string;
  model?: string;
}

function resolveModelId(engine: AGTEngine, explicit?: string): string {
  if (explicit) return explicit;
  if (typeof process !== "undefined") {
    if (process.env.SWT3_MODEL_ID) return process.env.SWT3_MODEL_ID;
    if (process.env.AGT_MODEL_ID) return process.env.AGT_MODEL_ID;
  }
  if (engine.model) return engine.model;
  if (engine.name) return `agt-${engine.name}`;
  return "agt-policy-engine";
}

function stringifyValue(val: unknown): string {
  if (val === null || val === undefined) return "";
  if (typeof val === "string") return val;
  try {
    return JSON.stringify(val);
  } catch {
    return String(val);
  }
}

interface GuardrailInfo {
  active: number;
  names: string[];
  passed: boolean;
}

function extractGuardrails(result: unknown): GuardrailInfo {
  const info: GuardrailInfo = { active: 0, names: [], passed: true };

  if (result && typeof result === "object" && !Array.isArray(result)) {
    const r = result as Record<string, unknown>;
    const policies = (r.policies_evaluated ?? r.guardrails) as unknown[] | undefined;
    if (Array.isArray(policies)) {
      info.active = policies.length;
      for (const p of policies) {
        if (p && typeof p === "object") {
          const pol = p as Record<string, unknown>;
          const name = (pol.name ?? pol.policy_name) as string | undefined;
          if (name) info.names.push(String(name));
          if (pol.result === "fail" || pol.passed === false) info.passed = false;
        } else if (typeof p === "string") {
          info.names.push(p);
        }
      }
    }
    const verdict = (r.verdict ?? r.decision) as string | undefined;
    if (typeof verdict === "string" && ["deny", "block", "fail", "rejected"].includes(verdict.toLowerCase())) {
      info.passed = false;
    }
  }

  return info;
}

function wrapMethod<T extends AGTEngine>(
  target: T,
  methodName: "evaluate" | "assess",
  witness: Witness,
  mid: string,
): void {
  const original = (target as any)[methodName];
  if (typeof original !== "function") return;
  const bound = original.bind(target);

  (target as any)[methodName] = (input: unknown, ...args: unknown[]): unknown => {
    const start = performance.now();
    const result = bound(input, ...args);

    const finish = (res: unknown): unknown => {
      const elapsed = Math.round(performance.now() - start);
      const guardrails = extractGuardrails(res);
      const record: InferenceRecord = {
        modelId: mid,
        modelHash: sha256Truncated(mid),
        promptHash: sha256Truncated(stringifyValue(input)),
        responseHash: sha256Truncated(stringifyValue(res)),
        latencyMs: elapsed,
        inputTokens: 0,
        outputTokens: 0,
        guardrailsActive: guardrails.active,
        guardrailsRequired: guardrails.active,
        guardrailPassed: guardrails.passed,
        hasRefusal: !guardrails.passed,
        provider: "microsoft-agt",
        guardrailNames: guardrails.names,
      };
      witness.record(record);
      return res;
    };

    if (result && typeof (result as any).then === "function") {
      return (result as Promise<unknown>).then(finish);
    }
    return finish(result);
  };
}

export function wrapAGT<T extends AGTEngine>(
  engine: T,
  witness: Witness,
  modelId?: string,
): T {
  const mid = resolveModelId(engine, modelId);
  const wrapped = Object.create(engine) as T;

  wrapMethod(wrapped, "evaluate", witness, mid);
  if (engine.assess) {
    wrapMethod(wrapped, "assess", witness, mid);
  }

  return wrapped;
}

File: packages/swt3-ai/swt3_ai/adapters/agt.py
"""SWT3 AI Witness SDK -- Microsoft AGT (Agent Governance Toolkit) Adapter.

Wraps any object with an evaluate() method (Microsoft AGT policy engine pattern),
minting witness anchors on each policy evaluation without modifying
the governance logic or adding framework dependencies.

SWT3 is the independent witness layer for AGT-managed agents. AGT enforces
policy; SWT3 proves what AGT decided, cryptographically, out-of-band.

Usage:

    from swt3_ai.adapters.agt import wrap_agt

    witnessed = wrap_agt(policy_engine, witness)
    decision = witnessed.evaluate(prompt, context={"model": "gpt-4o"})

Configuration:
    1. SWT3_DSN=https://axm_live_xxx@sovereign.tenova.io/MY_ENCLAVE
    2. SWT3_ENDPOINT + SWT3_API_KEY + SWT3_TENANT_ID
    3. wrap_agt(engine, witness=my_witness)

Duck-typed: works with any object that has an evaluate() method.
No Microsoft SDK import required.

Copyright (c) 2026 Tenable Nova LLC. Apache 2.0. Patent pending.
"""

from __future__ import annotations

import json
import logging
import os
import time
from typing import Any, Optional, TYPE_CHECKING

from ..fingerprint import sha256_truncated
from ..types import InferenceRecord

if TYPE_CHECKING:
    from ..witness import Witness

logger = logging.getLogger("swt3_ai.agt")


def _resolve_model_id(engine: Any, explicit: Optional[str] = None) -> str:
    if explicit:
        return explicit
    env_id = os.environ.get("SWT3_MODEL_ID") or os.environ.get("AGT_MODEL_ID")
    if env_id:
        return env_id
    if hasattr(engine, "model") and engine.model:
        return str(engine.model)
    if hasattr(engine, "name") and engine.name:
        return f"agt-{engine.name}"
    return "agt-policy-engine"


def _stringify(val: Any) -> str:
    if val is None:
        return ""
    if isinstance(val, str):
        return val
    try:
        return json.dumps(val, default=str)
    except Exception:
        return str(val)


def _extract_guardrails(result: Any) -> tuple:
    """Extract guardrail metadata from an AGT Decision BOM if available."""
    guardrails_active = 0
    guardrail_names: list = []
    guardrail_passed = True

    if isinstance(result, dict):
        policies = result.get("policies_evaluated") or result.get("guardrails") or []
        if isinstance(policies, (list, tuple)):
            guardrails_active = len(policies)
            for p in policies:
                if isinstance(p, dict):
                    name = p.get("name") or p.get("policy_name") or ""
                    if name:
                        guardrail_names.append(str(name))
                    if p.get("result") == "fail" or p.get("passed") is False:
                        guardrail_passed = False
                elif isinstance(p, str):
                    guardrail_names.append(p)
        verdict = result.get("verdict") or result.get("decision")
        if isinstance(verdict, str) and verdict.lower() in ("deny", "block", "fail", "rejected"):
            guardrail_passed = False
    elif hasattr(result, "policies_evaluated"):
        policies = getattr(result, "policies_evaluated", [])
        if isinstance(policies, (list, tuple)):
            guardrails_active = len(policies)

    return guardrails_active, guardrail_names, guardrail_passed


def wrap_agt(
    engine: Any,
    witness: Optional["Witness"] = None,
    model_id: Optional[str] = None,
    **overrides: Any,
) -> Any:
    """Wrap a Microsoft AGT policy engine to witness every evaluate() call.

    Args:
        engine: Any object with an evaluate() method (AGT engine or compatible).
        witness: Explicit Witness instance. If None, auto-creates from env.
        model_id: Model identifier override.
        **overrides: Override config (clearing_level, agent_id, etc.)

    Returns:
        Proxy object that behaves identically to the original engine
        but mints a witness anchor on each evaluation.
    """
    if witness is None:
        try:
            from ..witness import SWT3Witness
            dsn = os.environ.get("SWT3_DSN")
            if dsn:
                witness = SWT3Witness.from_dsn(dsn, **overrides)
            else:
                endpoint = os.environ.get("SWT3_ENDPOINT", "")
                api_key = os.environ.get("SWT3_API_KEY", "")
                tenant_id = os.environ.get("SWT3_TENANT_ID", "")
                if endpoint and api_key and tenant_id:
                    witness = SWT3Witness(
                        endpoint=endpoint,
                        api_key=api_key,
                        tenant_id=tenant_id,
                        **overrides,
                    )
        except Exception:
            pass

    if witness is None:
        logger.debug("No witness configured; AGT adapter is a no-op")
        return engine

    mid = _resolve_model_id(engine, model_id)

    class AGTProxy:
        """Transparent proxy that witnesses Microsoft AGT policy evaluations."""

        __slots__ = ("__target", "__witness")

        def __init__(self, target: Any, w: "Witness"):
            object.__setattr__(self, "_AGTProxy__target", target)
            object.__setattr__(self, "_AGTProxy__witness", w)

        def _record(self, prompt: Any, result: Any, elapsed: int) -> None:
            guardrails_active, guardrail_names, guardrail_passed = _extract_guardrails(result)
            record = InferenceRecord(
                model_id=mid,
                model_hash=sha256_truncated(mid),
                prompt_hash=sha256_truncated(_stringify(prompt)),
                response_hash=sha256_truncated(_stringify(result)),
                latency_ms=elapsed,
                input_tokens=0,
                output_tokens=0,
                guardrails_active=guardrails_active,
                guardrails_required=guardrails_active,
                guardrail_passed=guardrail_passed,
                has_refusal=not guardrail_passed,
                provider="microsoft-agt",
                guardrail_names=guardrail_names,
            )
            self.__witness.record(record)

        def evaluate(self, prompt: Any, *args: Any, **kwargs: Any) -> Any:
            start = time.perf_counter()
            result = self.__target.evaluate(prompt, *args, **kwargs)
            elapsed = round((time.perf_counter() - start) * 1000)
            self._record(prompt, result, elapsed)
            return result

        def assess(self, config: Any, *args: Any, **kwargs: Any) -> Any:
            if not hasattr(self.__target, "assess"):
                raise AttributeError("assess")
            start = time.perf_counter()
            result = self.__target.assess(config, *args, **kwargs)
            elapsed = round((time.perf_counter() - start) * 1000)
            self._record(config, result, elapsed)
            return result

        def __getattr__(self, name: str) -> Any:
            if name.startswith("_"):
                raise AttributeError(name)
            return getattr(self.__target, name)

    return AGTProxy(engine, witness)
